### PR TITLE
Document All The Things

### PR DESCRIPTION
**Pull request by CodaFi**
- [x] Break up large sections with `Mark`s.
- [x] Clarify pipes terminology (`Producer`, `Consumer`, `Pipe`, `Effect`, `Client`, `Server`).
- [x] Document Operators properly.
- [x] Document Group Producer properly.
- [x] Diagrams, Diagrams, Diagrams

### CODE REVIEW

**Comment by CodaFi** [comment]
I'm unclear on a few of the operators here.  Specifically

`|<|`
and 
`>|>`

**Comment by pthariensflame** [comment]
@CodaFi Those could probably stand to be renamed.  :P

The correspondence between operator names in `Aquifer` and `pipes` is as close as I could possibly make it, but some of the [original `pipes` operators](http://hackage.haskell.org/package/pipes/docs/doc-index.html) (specifically, those containing `\` or `/` characters) had to be renamed due to limitations of Swift syntax.

**Comment by CodaFi** [comment]
Ja, that I totally understand.  I was wondering, specifically, _which_ operators they mapped to.

File: Aquifer/Auxiliary.swift [hunk 1]
```diff
@@ -8,35 +8,28 @@
...
 
-private func leftInner<A, B, C>() -> Proxy<(), Either<A, C>, (), Either<B, C>, A> {
-    return await() >>- {
-        switch $0 {
-        case let .Left(x): return pure(x)
-        case let .Right(y): return yield(Either.Right(y)) >>- { _ in leftInner() }
-        }
-    }
-}
-
```
Comment (pthariensflame) [inline Aquifer/Auxiliary.swift:26]: Why was `leftInner` deleted?  Same question for `rightInner` below.

EDIT:  Never mind; I see what you've done here.

Comment (CodaFi) [inline Aquifer/Auxiliary.swift:26]: Nope, nothing got removed.  They were moved to the bottom of the file.

Comment (pthariensflame) [inline Aquifer/Auxiliary.swift:26]: @CodaFi Understood.  :)

**Comment by pthariensflame** [comment]
I just went through the code and I actually can't figure out the some of the operator name correspondences myself, which is, I guess, yet more evidence of how terrible they are.  I swear they made sense to me at the time!  :(

**Comment by pthariensflame** [comment]
Any suggestions for a better name correspondence?  We need the following operators to have Swift-ier names (each pair should remain a pair):
- `//>` and `<\\`
- `>\\` and `//<`
- `/>/` and `\<\`
- `\>\` and `/</`

**Comment by CodaFi** [comment]
I think you did a fine job with the first two just converting the slashes to pipes.  The last couple I'm not sure of... Swift won't take any of those, and there's a duplication problem if you just convert to `|>|` and`|<|`.  Unless you're cool with overloading here?

**Comment by pthariensflame** [comment]
I would rather not overload here, even if Swift allows it, because it would be confusing for the user as to which overload of each operator applied where.  Both of those pairs are used on `Proxy` Kleisli morphisms, so it would be very hard to tell what a given piece of code did if they were overloaded.

File: Aquifer/Proxy.swift [hunk 2]
```diff
@@ -8,90 +8,44 @@
...
+///
+///     /// Sends requests of type `RequestT` and recieves responses of type `RespondT`.
+///     ///
+///     /// `Client`s only request and never respond.
+///     typealias Client<RequestT, RespondT, Result> = Proxy<RequestT, RespondT, (), X, Result>
+///
+///     /// Receives values of type `ReceiveT` and responds with values of type `RespondT`.
+///     ///
+///     /// `Server`s only respond and never request.
+///     typealias Server<ReceiveT, RespondT, Result> = Proxy<X, (), ReceiveT, RespondT, Result>
```
Comment (CodaFi) [inline Aquifer/Proxy.swift:46]: How do you want to have this show up in documentation since we can't do any of this polymorphic marker type stuff?

Comment (pthariensflame) [inline Aquifer/Proxy.swift:46]: I think the way you have it there is fine for now.  It means that if Swift ever _does_ get generic `typealiases`, we can simply transfer them into actual code from the comments.

Comment (CodaFi) [inline Aquifer/Proxy.swift:46]: Sorry, should have been more specific.  I mean when I went through and documented everything I used the word `pipe` everywhere, and by the end of it I felt dirty.  I mean, how do we let people know which of the 6 proxies they're dealing with in a clear way in the docs for each function?

Comment (pthariensflame) [inline Aquifer/Proxy.swift:46]: The functions are mostly usage-unspecific, and for good reason.

Comment (pthariensflame) [inline Aquifer/Proxy.swift:46]: That being said, you can look at their counterparts from the original `pipes` and see what type aliases get used where in there.

File: Aquifer/Auxiliary.swift [hunk 3]
```diff
@@ -46,22 +39,54 @@ associativity left
...
     return scan1(stepWith: step, initializeWith: identity, extractWith: identity)
 }
 
+/// Yields a pipe that produces left-scanned values with the given step function.  The pipe is not
+/// required to have an initial value, but one is expected to be produced by the `initial` function.
 public func scan1<A, UI, DO, FR>(stepWith step: (A, UI) -> A, initializeWith initial: UI -> A, extractWith extractor: A -> DO) -> Proxy<(), UI, (), DO, FR> {
     return await() >>- { scan(stepWith: step, initializeWith: initial($0), extractWith: extractor) }
 }
+
+/// Implementation Details Follow
```
Comment (pthariensflame) [inline Aquifer/Auxiliary.swift]: Why is this a doc comment.  Genuinely curious, since I can't see a reason for it.

Comment (CodaFi) [inline Aquifer/Auxiliary.swift]: I did this with SwiftCheck mostly for me.  I can get rid of them all if you'd prefer.

Comment (pthariensflame) [inline Aquifer/Auxiliary.swift]: No, they're fine.  You'll be working on this codebase too, and I don't mind them, so if they help you then go for it!  :)

File: Aquifer/Proxy.swift (hunk 2 above)
Comment (CodaFi) [inline Aquifer/Proxy.swift:46]: Mhm, but I'm looking for a style we can use in the docs consistently.  For example

``` swift
/// Converts a given sequence into a pipe that produces elements of the same.
public func each<UO, UI, S: SequenceType>(seq: S) -> Proxy<UO, UI, (), S.Generator.Element, ()> {
    return Proxy(eachRepr(seq.generate()))
}
```

Can become

``` swift
/// Converts a given sequence into a `Producer` of the elements of the sequence.
public func each<UO, UI, S: SequenceType>(seq: S) -> Proxy<UO, UI, (), S.Generator.Element, ()> {
    return Proxy(eachRepr(seq.generate()))
}
```

or 

``` swift
/// Converts a given sequence into a pipe that produces elements of the same.
///
/// Yields a Producer<S.Generator.Element, ()>
public func each<UO, UI, S: SequenceType>(seq: S) -> Proxy<UO, UI, (), S.Generator.Element, ()> {
    return Proxy(eachRepr(seq.generate()))
}
```

File: Aquifer/Auxiliary.swift (hunk 3 above)
Comment (CodaFi) [inline Aquifer/Auxiliary.swift]: They might be better as `MARK: -`s

File: Aquifer/Proxy.swift (hunk 2 above)
Comment (pthariensflame) [inline Aquifer/Proxy.swift:46]: I prefer the former of the two you gave, although if we could combine the two in a way that wasn't redundant then that would be ideal.

File: Aquifer/Auxiliary.swift (hunk 3 above)
Comment (pthariensflame) [inline Aquifer/Auxiliary.swift]: Now that you mention it, yeah, they would.  Do you want to make the same change in `SwiftCheck` (and elsewhere)?

Comment (CodaFi) [inline Aquifer/Auxiliary.swift]: Yeah, totally!  I'll draw that one up after we're done here.

File: Aquifer/Proxy.swift (hunk 2 above)
Comment (CodaFi) [inline Aquifer/Proxy.swift:46]:
```
/// Converts a given sequence into a `Producer<S.Generator.Element, ()>` of the elements of the sequence.
```

File: Aquifer/Auxiliary.swift (hunk 3 above)
Comment (pthariensflame) [inline Aquifer/Auxiliary.swift]: :+1:

**Comment by CodaFi** [comment]
I'm happy with this for now.  Capping changes here.

Ready for review.

File: Aquifer/Proxy.swift (hunk 2 above)
Comment (pthariensflame) [inline Aquifer/Proxy.swift:46]: :+1:

File: Aquifer/Basic.swift [hunk 4]
```diff
@@ -8,212 +8,151 @@
...
+    return for_(cat(), each)
 }
 
+
+/// Returns whether all elements of the receiver satisfy a given predicate.
 public func all<V>(p: Proxy<X, (), (), V, ()>, predicate: V -> Bool) -> Bool {
     return isEmpty(p >-> filter { !predicate($0) })
 }
 
+/// Returns whether any element of the receiver satisfy a given predicate.
```
Comment (pthariensflame) [inline Aquifer/Basic.swift]: Grammar!  :P

File: Aquifer/Grouping.swift [hunk 5]
```diff
@@ -8,36 +8,14 @@
...
-        }
-    }
-}
-
+/// A producer that uses the Free Monad Transformer (unrepresentable in Swift) to delimit groupings
+/// of produced values.
+///
+/// Ideally:
+///
+///    typealias GroupedProducer<V, R> = Free<Producer<V> R>
```
Comment (pthariensflame) [inline Aquifer/Grouping.swift]: Should probably be `Free<Producer<V, $0>, R>` just for consistency with Swift syntax.  :)

File: Aquifer/Grouping.swift [hunk 6]
```diff
@@ -50,18 +28,26 @@ public struct GroupedProducer<V, R> {
...
 }
 
+/// Forces a `GroupedProducer` to evaluate lazily.
 public func delay<V, R>(@autoclosure(escaping) p: () -> GroupedProducer<V, R>) -> GroupedProducer<V, R> {
     return GroupedProducer(p().repr)
 }
 
+/// Pull the first value out of the given `GroupedProducer`.
+///
+/// If the subsequent state of the `Producer` is `.End` or terminated, the result is `.Left`
```
Comment (pthariensflame) [inline Aquifer/Grouping.swift]: `.End` is implementation detail and should probably not be mentioned.

File: Aquifer/Proxy.swift [hunk 7]
```diff
@@ -8,96 +8,66 @@
...
+///     /// A computation that can await values of type `A` and yield values of type `B`.
+///     ///
+///     /// `Pipe`s can both await and yield.
+///     typealias Pipe<A, B, Result> = Proxy<(), A, (), B, Result>
+///
+///     /// A computation that can await values of type `A`.
+///     ///
+///     /// `Consumer`s can only await.
+///     typealias Consumer<A, Result> = Proxy<(), A, (), X, Result>
+///
```
Comment (pthariensflame) [inline Aquifer/Proxy.swift:37]: Why is there an extra space here?

File: Aquifer/Proxy.swift [hunk 8]
```diff
@@ -8,96 +8,66 @@
...
 /// UO - upstream   output
 /// UI - upstream   input
 /// DI — downstream input
 /// DO — downstream output
 /// FR — final      result
+///
+/// Which can be represented diagrammatically by:
+///
+/// Diagrammatically, you can think of a `Proxy` as having the following shape:
```
Comment (pthariensflame) [inline Aquifer/Proxy.swift]: This sentence and the one before should be combined for clarity and grammar.

File: Aquifer/Simple.swift [hunk 9]
```diff
@@ -19,38 +22,43 @@ public func next<DO, FR>(p: Proxy<X, (), (), DO, FR>) -> Either<FR, (DO, Proxy<X
     }
 }
 
+/// Discards the value it is given.
+///
+/// The resulting pipe responds to all requests with `()`.
```
Comment (pthariensflame) [inline Aquifer/Simple.swift]: This isn't quite accurate.  In Haskell, this is `const $ return ()`.

File: Aquifer/Simple.swift [hunk 10]
```diff
@@ -8,9 +8,12 @@
 
 // roughly `Pipes`
 
-import Foundation
 import Swiftz
 
+/// Pull the first value out of the given pipe.
+///
+/// If the subsequent state of the pipe is `.Pure` or terminated, the result is `.Left` containing
```
Comment (pthariensflame) [inline Aquifer/Simple.swift]: `.Pure is implementation detail and should probably not be mentioned.

File: Aquifer/Simple.swift [hunk 11]
```diff
@@ -8,9 +8,12 @@
...
+/// the value.  Else the result is `.Right` containing the value and the next state of the pipe.
```
Comment (pthariensflame) [inline Aquifer/Simple.swift]: "Otherwise" rather than "else," here.

File: Aquifer/Simple.swift [hunk 12]
```diff
@@ -19,38 +22,43 @@ public func next<DO, FR>(p: Proxy<X, (), (), DO, FR>) -> Either<FR, (DO, Proxy<X
...
 }
 
+/// The identity `Pipe`.
+///
+/// Like the Unix `cat` program, pushes any given input as output without modification.
 public func cat<DT, FR>() -> Proxy<(), DT, (), DT, FR> {
     return pull(())
 }
 
+/// For each value `yield`ed in the given pipe and replaces it with the result of applying the
```
Comment (pthariensflame) [inline Aquifer/Simple.swift]: Work on grammar here.

**Comment by pthariensflame** [comment]
Other than the mentioned nitpicks, this looks good!

**Comment by CodaFi** [comment]
There, that should do it.

File: Aquifer/Basic.swift (hunk 4 above)
Comment (pthariensflame) [inline Aquifer/Basic.swift]: This still needs fixing.

**Comment by pthariensflame** [comment]
You missed a single thing.  :)

File: Aquifer/Grouping.swift (hunk 5 above)
Comment (pthariensflame) [inline Aquifer/Grouping.swift]: I'd like some discussion on this.

Comment (CodaFi) [inline Aquifer/Grouping.swift]: Naw, I cave.

File: Aquifer/Proxy.swift (hunk 7 above)
Comment (CodaFi) [inline Aquifer/Proxy.swift:37]: An unfortunate side effect of the way this was documented.  It's fixed now.

**Comment by CodaFi** [comment]
Not anymore :sparkle:

**Comment by CodaFi** [comment]
終わりますか？

**Comment by pthariensflame** [comment]
Sí, ahora esta bien.  :)

**Comment by CodaFi** [comment]
凄い！

You have the bridge

**Comment by pthariensflame** [comment]
Thank you, First.

**Comment by CodaFi** [comment]
:sparkles:

**Comment by pthariensflame** [comment]
:heart:

**Pull request merged**